### PR TITLE
2328/address verification

### PR DESCRIPTION
**Pull request by jaredcwhite**
## Issue Overview

This PR addresses #2328 

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

This PR adds one extra step to the public Application > Primary Applicant Address form to validate and confirm the address via the MapBox API.

## How Can This Be Tested/Reviewed?

Try filling out an application for a listing on the Public site, and on the second "contact" step try putting in both real addresses and garbled data to make sure the interface is reasonably robust to allow saving the new address (or at least what was entered). Also: https://github.com/bloom-housing/bloom/pull/2412#issuecomment-1015834594

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

**Comment by jaredcwhite** [comment]
A couple of notes:

* The partners test is failing due to another issue that (is? was?) on dev, not affected by this PR.
* The git diff in the "Files changed" is really weird, most of the stuff that looks changed isn't actually, it just got added to a conditional block. Subsequently I'm pretty sure most if not all of the accessibility warnings are bogus.

**Comment by dominikx96** [comment]
@jaredcwhite this works good to me, we just need to remember about this before merging: https://github.com/bloom-housing/bloom/issues/2328#issuecomment-1015830050

The second question is - when a user selects an option generated by Mapbox geocoder, maybe we should also save latitude and longitude?

**Review by YazeedLoonat — CHANGES_REQUESTED** [review]
Looks good to me, just a question on here, and can you please add QA notes to the issue

File: sites/public/src/forms/applications/ValidateAddress.tsx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  })
+
+  geocodingClient
+    .forwardGeocode({
+      query: `${address.street}, ${address.city}, ${address.state}, ${address.zipCode}`,
+      limit: 1,
+    })
+    .send()
+    .then((response) => {
+      const [street, city, region] = response.body.features[0].place_name.split(", ")
```
[the remark this answers is not in the document]
Comment (jaredcwhite) [reply sites/public/src/forms/applications/ValidateAddress.tsx:26]: I tried a wide variety of inputs including completely garbage text, and the most it ever came back funky I was able to detect due to a missing Zip code (which is is part of that code block), so I think we're OK here. Their "fuzzy search" is indeed very fuzzy 😄

**Comment by jaredcwhite** [comment]
> The second question is - when a user selects an option generated by Mapbox geocoder, maybe we should also save latitude and longitude?

@dominikx96 I'm not sure, I don't know if we need lat/lng for applications. Isn't that just for listings?

**Comment by seanmalbert** [comment]
> > The second question is - when a user selects an option generated by Mapbox geocoder, maybe we should also save latitude and longitude?
> 
> @dominikx96 I'm not sure, I don't know if we need lat/lng for applications. Isn't that just for listings?

I do not think we need to save the coordinates of the applicants address.

**Review by seanmalbert — CHANGES_REQUESTED** [review]
@jaredcwhite ,
This looks good. Can you add in a test with garbage address info, that way if the API becomes less fuzzy for whatever reason, we'll be able to catch it sooner. I was also wondering about if it made sense to set the country to always be the U.S. for the API call. I think this might cut out a lot of the fuzziness we're seeing.

**Comment by seanmalbert** [comment]
@jaredcwhite ,
I asked @kathyccheng about always setting US as the country when calling the API and she thinks that's a viable option. Can you update this to be the case along with a test or two with garbage data.

**Comment by jaredcwhite** [comment]
@seanmalbert Sounds good, I'll add the country in the search query code, as well as see if I can expand the E2E test with a couple rounds of varied data entry.

**Comment by jaredcwhite** [comment]
OK, I discussed with Jesse and now the error state looks better (see https://github.com/bloom-housing/bloom/issues/2328#issuecomment-1021667122)…think this is ready to merge now.

**Review by seanmalbert — APPROVED** [review]

**Pull request merged**